packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/SemanticProjectImportParticipant.java

```diff
@@ -0,0 +1,220 @@
+
+    private static final String DOCUMENTS_FOLDER = "documents";
+
+    private static final String ZIP_FOLDER_SEPARATOR = "/";
+
+    private final IEditingContextSearchService editingContextSearchService;
+
+    private final IUploadFileLoader uploadDocumentLoader;
+
+    private final List<IUploadDocumentReportProvider> uploadDocumentReportProviders;
+
+    private final IEditingContextPersistenceService editingContextPersistenceService;
+
+    private final IRewriteProxiesService rewriteProxiesService;
+
+    public SemanticProjectImportParticipant(
+            IEditingContextSearchService editingContextSearchService, IUploadFileLoader uploadDocumentLoader, List<IUploadDocumentReportProvider> uploadDocumentReportProviders,
+            IEditingContextPersistenceService editingContextPersistenceService, IRewriteProxiesService rewriteProxiesService) {
+        this.editingContextSearchService = Objects.requireNonNull(editingContextSearchService);
+        this.uploadDocumentLoader = Objects.requireNonNull(uploadDocumentLoader);
+        this.uploadDocumentReportProviders = Objects.requireNonNull(uploadDocumentReportProviders);
+        this.editingContextPersistenceService = Objects.requireNonNull(editingContextPersistenceService);
+        this.rewriteProxiesService = Objects.requireNonNull(rewriteProxiesService);
+    }
+
+    @Override
+    public boolean canHandle(IDomainEvent event) {
+        return event instanceof ProjectSemanticDataCreatedEvent projectSemanticDataCreatedEvent
+                && projectSemanticDataCreatedEvent.causedBy() instanceof SemanticDataCreatedEvent semanticDataCreatedEvent
+                && semanticDataCreatedEvent.causedBy() instanceof ProjectCreatedEvent projectCreatedEvent
+                && projectCreatedEvent.causedBy() instanceof InitProjectContentInput;
+
+    }
+
+    @Override
+    public ImportContentStatus handle(IDomainEvent event, IProjectBinaryContent projectContent) {
+        ProjectSemanticDataCreatedEvent projectSemCreationEvent = (ProjectSemanticDataCreatedEvent) event;
+        SemanticDataCreatedEvent semCreationEvent = (SemanticDataCreatedEvent) projectSemCreationEvent.causedBy();
+        InitProjectContentInput originalInput = (InitProjectContentInput) event.causedBy().causedBy().causedBy();
+
+        var documentsToUpload = getDocuments(projectContent);
+        if (!documentsToUpload.isEmpty()) {
+            return copySemantic(documentsToUpload, semCreationEvent.semanticData().getId().toString(), originalInput);
+        }
+        return new ImportContentStatus(true, List.of());
+    }
+
+    /**
+     * Gets the document to upload from the {@link IProjectBinaryContent}.
+     *
+     * @param projectUploadContent
+     *         the project content
+     * @return a map linking document id with their matching {@link UploadFile}
+     */
+    private List<Pair<String, UploadFile>> getDocuments(IProjectBinaryContent projectUploadContent) {
```

Comment:
We will not in any way start depending on internal data structure from Spring just to create our owns without any link to their concerns.

Reply:
This is just a plain old Map anyway